ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/BaseDateKudaf.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udaf;
+
+import io.confluent.ksql.GenericKey;
+import io.confluent.ksql.function.BaseAggregateFunction;
+import io.confluent.ksql.function.ParameterInfo;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import java.util.Collections;
+import java.util.Date;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import org.apache.kafka.streams.kstream.Merger;
+
+public class BaseDateKudaf<T extends Date> extends BaseAggregateFunction<T, T, T> {
+
+  private final BiFunction<T, T, T> aggregatePrimitive;
+
+  private final BiFunction<T, T, T> maxAggregatePrimitive = (a, b) -> {
+    if (a.after(b)) {
```

Comment:
I think there's a possible refactoring which would let us get rid of a ton of classes here.

All of the types involved are `Comparable`.  That ordering ought to be sufficient to calculate mins/maxs.  (E.g., we don't need to use `after`.)

One option would be to replace `BaseNumberKudaf` with a `ComparableKudaf`, create concrete Min and Max versions, and then in the `M*AggFunctionFactory`, call the constructor for the concrete classes with the parameters to make it go.

The second option to consider would be trying to switch these comparable functions to the UdafFactory method approach (set `collect_list` and `latest_by_offset` for examples).

Reply:
Great ideas! Let me have another go to make this cleaner ;)